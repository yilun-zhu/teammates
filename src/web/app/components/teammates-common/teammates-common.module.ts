import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { EnumToArrayPipe } from './enum-to-array.pipe';
import { FormatDateDetailPipe } from './format-date-detail.pipe';
import { InstructorRoleNamePipe } from './instructor-role-name.pipe';
import { PublishStatusNamePipe } from './publish-status-name.pipe';
import { QuestionTypeNamePipe } from './question-type-name.pipe';
import { SafeHtmlPipe } from './safe-html.pipe';
import { SubmissionStatusNamePipe } from './submission-status-name.pipe';

/**
 * Common module in the project.
 */
@NgModule({
  imports: [
    CommonModule,
  ],
  declarations: [
    EnumToArrayPipe,
    SubmissionStatusNamePipe,
    PublishStatusNamePipe,
    FormatDateDetailPipe,
    SafeHtmlPipe,
    QuestionTypeNamePipe,
    InstructorRoleNamePipe,
  ],
  exports: [
    EnumToArrayPipe,
    SubmissionStatusNamePipe,
    PublishStatusNamePipe,
    FormatDateDetailPipe,
    SafeHtmlPipe,
    QuestionTypeNamePipe,
    InstructorRoleNamePipe,
  ],
})
export class TeammatesCommonModule { }
