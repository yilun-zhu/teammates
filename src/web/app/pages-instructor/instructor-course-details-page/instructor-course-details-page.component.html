<div>
  <h1>Course Details</h1>
  <div class="card">
    <div class="card-body fill-plain">
      <div class="form">
        <div class="form-group row">
          <label class="col-sm-3 text-right">Course ID:</label>
          <div class="col-sm-6">
            <p class="form-control-static">{{courseDetails.course.courseId}}</p>
          </div>
        </div>
        <div class="form-group row">
          <label class="col-sm-3 text-right">Course name:</label>
          <div class="col-sm-6" id="coursename">
            <p class="form-control-static">{{courseDetails.course.courseName}}</p>
          </div>
        </div>
        <div class="form-group row">
          <label class="col-sm-3 text-right">Sections:</label>
          <div class="col-sm-6">
            <p class="form-control-static">{{courseDetails.stats.sectionsTotal}}</p>
          </div>
        </div>
        <div class="form-group row">
          <label class="col-sm-3 text-right">Teams:</label>
          <div class="col-sm-6">
            <p class="form-control-static">{{courseDetails.stats.teamsTotal}}</p>
          </div>
        </div>
        <div class="form-group row">
          <label class="col-sm-3 text-right">Total students:</label>
          <div class="col-sm-6">
            <p class="form-control-static">{{courseDetails.stats.studentsTotal}}</p>
          </div>
        </div>
      </div>
      <div class="form-group row">
        <label class="col-sm-3 control-label text-right">Instructors:</label>
        <div class="col-sm-6">
          <div *ngFor="let instructor of instructors">
            {{instructor.role | instructorRoleName}}: {{instructor.name}} ({{instructor.email}})
          </div>
        </div>
      </div>
      <div *ngIf="sections.length > 0">
        <div class="form-group">
          <div class="text-center">
            <button type="button" class="btn btn-primary" (click)="openModal(remindStudentsModal)" value="Remind Students to Join">
              <i class="fa fa-envelope"></i>
              Remind Students to Join
            </button>
            <button type="button" class="btn btn-primary" (click)="downloadAllStudentsFromCourse(courseDetails.course.courseId)">
              <i class="fa fa-download"></i>
              Download Student List
            </button>
            <button type="button" class="btn btn-danger" (click)="openModal(deleteAllStudentsModal)" value="Delete All Students">
              <i class="fa fa-trash"></i>
              Delete All Students
            </button>
            <div>
              <span>Non-English characters not displayed properly in the downloaded file?
                <span class="btn-link" (click)="openModal(studentTableModal); loadStudentsListCsv(courseDetails.course.courseId)">
                  Click here.
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div *ngIf="sections.length === 0, else studentListTable" class="alert alert-warning">
    <i class="fa fa-exclamation-circle text-muted"></i>
    There are no students in this course. Click
    <a routerLink="/web/instructor/courses/enroll" [queryParams]="{courseid: courseDetails.course.courseId}">here</a> to enroll students.
  </div>
  <ng-template #studentListTable class="card border border-primary mb-4">
    <div class="card-body p-0">
      <div class="card-header text-dark-blue bg-light-blue font-weight-bold">[{{courseDetails.course.courseId}}]</div>
      <div class="card-body p-0">
        <tm-student-list [courseId]="courseDetails.course.courseId" [sections]="sections" [useGrayHeading]="true" [enableRemindButton]="true">
        </tm-student-list>
      </div>
    </div>
  </ng-template>
  <ng-template #remindStudentsModal let-modal>
    <div class="modal-header alert-info">
      <i class="fa fa-info-circle fa-2x"></i>
      <h4 class="modal-title">Confirm sending join requests</h4>
      <button type="button" class="close" aria-label="Close" (click)="modal.dismiss()">&times;</button>
    </div>
    <div class="modal-body">
      Usually, there is no need to use this feature because TEAMMATES sends an automatic invite to students at the opening
      time of each session. Send a join request to all yet-to-join students in {{courseDetails.course.courseId}} anyway?
      <br>
    </div>
    <div class="modal-footer">
      <button type="button" class="modal-btn-cancel btn-lg btn-default" (click)="modal.dismiss()">No, cancel the operation</button>
      <button type="button" class="modal-btn-ok btn-lg btn-danger" (click)="remindAllStudentsFromCourse(courseDetails.course.courseId); modal.dismiss()">Yes</button>
    </div>
  </ng-template>
  <ng-template #deleteAllStudentsModal let-modal>
    <div class="modal-header alert-danger">
      <i class="fa fa-times-circle fa-2x"></i>
      <h4 class="modal-title">Confirm Deletion</h4>
      <button type="button" class="close" aria-label="Close" (click)="modal.dismiss()">&times;</button>
    </div>
    <div class="modal-body">
      Are you sure you want to remove all students from the course {{courseDetails.course.courseId}}?
      <br>
    </div>
    <div class="modal-footer">
      <button type="button" class="modal-btn-cancel btn-lg btn-default" (click)="modal.dismiss()">No, cancel the operation</button>
      <button type="button" class="modal-btn-ok btn-lg btn-danger" (click)="deleteAllStudentsFromCourse(courseDetails.course.courseId); modal.dismiss()">Yes</button>
    </div>
  </ng-template>
  <ng-template #studentTableModal let-modal>
    <div class="modal-header">
      <span>
        Tips: <span class="btn-link" (click)="copyContent(courseStudentListAsCsv)">Click here</span>
        to copy the table, then <kbd>Ctrl + V</kbd> or <kbd>âŒ˜ + V</kbd> to PASTE to your Excel Workbook.
      </span>
      <button type="button" class="close" aria-label="Close" (click)="modal.dismiss('Cross click')">
        <span aria-hidden="true">&times;</span>
      </button>
    </div>
    <div class="modal-body">
      <tm-ajax-preload *ngIf="loading"></tm-ajax-preload>
      <div [innerHTML]="convertToHtmlTable(courseStudentListAsCsv)"></div>
      <b>{{ isAjaxSuccess ? '' : 'There was an error loading data. Please try again.' }}</b>
    </div>
  </ng-template>
</div>
