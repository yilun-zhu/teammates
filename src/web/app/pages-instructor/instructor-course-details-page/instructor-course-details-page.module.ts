import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { ClipboardModule } from 'ngx-clipboard';
import { AjaxPreloadModule } from '../../components/ajax-preload/ajax-preload.module';
import { TeammatesCommonModule } from '../../components/teammates-common/teammates-common.module';
import { StudentListModule } from '../student-list/student-list.module';
import { InstructorCourseDetailsPageComponent } from './instructor-course-details-page.component';

/**
 * Module for instructor course details page.
 */
@NgModule({
  declarations: [
    InstructorCourseDetailsPageComponent,
  ],
  exports: [
    InstructorCourseDetailsPageComponent,
  ],
  imports: [
    CommonModule,
    TeammatesCommonModule,
    ClipboardModule,
    RouterModule,
    StudentListModule,
    AjaxPreloadModule,
  ],
})
export class InstructorCourseDetailsPageModule { }
