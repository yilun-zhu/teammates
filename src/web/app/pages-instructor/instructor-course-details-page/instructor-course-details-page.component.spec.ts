import { HttpClientTestingModule } from '@angular/common/http/testing';
import { Component, Input } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { MatSnackBarModule } from '@angular/material';
import { RouterTestingModule } from '@angular/router/testing';
import { ClipboardModule } from 'ngx-clipboard';
import { Course, Instructor, InstructorPermissionRole, JoinState } from '../../../types/api-output';
import { TeammatesCommonModule } from '../../components/teammates-common/teammates-common.module';
import { InstructorCourseDetailsPageComponent } from './instructor-course-details-page.component';

@Component({ selector: 'tm-student-list', template: '' })
class StudentListStubComponent {
  @Input() courseId: string = '';
  @Input() useGrayHeading: boolean = true;
  @Input() sections: Object[] = [];
  @Input() enableRemindButton: boolean = true;
}
@Component({ selector: 'tm-ajax-preload', template: '' })
class AjaxPreloadComponent {}

const course: Course = {
  courseId: 'CS101',
  courseName: 'Introduction to CS',
  timeZone: '',
  creationTimestamp: 0,
  deletionTimestamp: 0,
};

const student: any = {
  name: 'Jamie',
  email: 'jamie@gmail.com',
  status: 'Yet to join',
  team: 'Team 1',
};

describe('InstructorCourseDetailsPageComponent', () => {
  let component: InstructorCourseDetailsPageComponent;
  let fixture: ComponentFixture<InstructorCourseDetailsPageComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        InstructorCourseDetailsPageComponent,
        StudentListStubComponent,
        AjaxPreloadComponent,
      ],
      imports: [
        HttpClientTestingModule,
        TeammatesCommonModule,
        RouterTestingModule,
        ClipboardModule,
        MatSnackBarModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(InstructorCourseDetailsPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with a course with one co-owner and no students, and populated course student list', () => {
    const stats: any = {
      sectionsTotal: 0,
      teamsTotal: 0,
      studentsTotal: 0,
    };
    const coOwner: Instructor = {
      courseId: course.courseId,
      joinState: JoinState.JOINED,
      googleId: 'Hodor',
      name: 'Hodor',
      email: 'hodor@gmail.com',
      role: InstructorPermissionRole.INSTRUCTOR_PERMISSION_ROLE_COOWNER,
      displayedToStudentsAs: 'Hodor',
      isDisplayedToStudents: true,
    };
    const courseDetails: any = {
      course,
      stats,
    };
    component.courseDetails = courseDetails;
    component.instructors = [coOwner];
    component.courseStudentListAsCsv = 'a,b';
    component.loading = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with a course with one co-owner and one student, and ajax failure', () => {
    const stats: any = {
      sectionsTotal: 1,
      teamsTotal: 1,
      studentsTotal: 1,
    };
    const coOwner: Instructor = {
      courseId: course.courseId,
      joinState: JoinState.JOINED,
      googleId: 'Bran',
      name: 'Bran',
      email: 'bran@gmail.com',
      role: InstructorPermissionRole.INSTRUCTOR_PERMISSION_ROLE_COOWNER,
      displayedToStudentsAs: 'Bran',
      isDisplayedToStudents: false,
    };
    const courseDetails: any = {
      course,
      stats,
    };
    const studentListSectionData: any = {
      sectionName: 'Tutorial Group 1',
      isAllowedToViewStudentInSection: true,
      isAllowedToModifyStudent: true,
      students: [student],
    };
    component.sections = [studentListSectionData];
    component.courseDetails = courseDetails;
    component.instructors = [coOwner];
    component.isAjaxSuccess = false;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
