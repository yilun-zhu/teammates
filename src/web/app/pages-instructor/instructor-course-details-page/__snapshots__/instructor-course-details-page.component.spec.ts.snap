// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructorCourseDetailsPageComponent should snap with a course with one co-owner and no students, and populated course student list 1`] = `
<tm-instructor-course-details-page
  clipboardService={[Function ClipboardService]}
  courseDetails={[Function Object]}
  courseService={[Function CourseService]}
  courseStudentListAsCsv={[Function String]}
  httpRequestService={[Function HttpRequestService]}
  instructors={[Function Array]}
  isAjaxSuccess={[Function Boolean]}
  loading={[Function Boolean]}
  navigationService={[Function NavigationService]}
  ngbModal={[Function NgbModal]}
  route={[Function ActivatedRoute]}
  router={[Function Router]}
  sections={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
>
  <div>
    <h1>
      Course Details
    </h1>
    <div
      class="card"
    >
      <div
        class="card-body fill-plain"
      >
        <div
          class="form"
        >
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Course ID:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                CS101
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Course name:
            </label>
            <div
              class="col-sm-6"
              id="coursename"
            >
              <p
                class="form-control-static"
              >
                Introduction to CS
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Sections:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                0
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Teams:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                0
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Total students:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                0
              </p>
            </div>
          </div>
        </div>
        <div
          class="form-group row"
        >
          <label
            class="col-sm-3 control-label text-right"
          >
            Instructors:
          </label>
          <div
            class="col-sm-6"
          >
            
            <div>
               Co-owner: Hodor (hodor@gmail.com) 
            </div>
          </div>
        </div>
        
      </div>
    </div>
    
    <div
      class="alert alert-warning"
    >
      <i
        class="fa fa-exclamation-circle text-muted"
      />
       There are no students in this course. Click 
      <a
        href="/web/instructor/courses/enroll?courseid=CS101"
        ng-reflect-query-params="[object Object]"
        ng-reflect-router-link="/web/instructor/courses/enroll"
        routerlink="/web/instructor/courses/enroll"
      >
        here
      </a>
       to enroll students. 
    </div>
    
    
    
    
  </div>
</tm-instructor-course-details-page>
`;

exports[`InstructorCourseDetailsPageComponent should snap with a course with one co-owner and one student, and ajax failure 1`] = `
<tm-instructor-course-details-page
  clipboardService={[Function ClipboardService]}
  courseDetails={[Function Object]}
  courseService={[Function CourseService]}
  courseStudentListAsCsv=""
  httpRequestService={[Function HttpRequestService]}
  instructors={[Function Array]}
  isAjaxSuccess="false"
  loading="false"
  navigationService={[Function NavigationService]}
  ngbModal={[Function NgbModal]}
  route={[Function ActivatedRoute]}
  router={[Function Router]}
  sections={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
>
  <div>
    <h1>
      Course Details
    </h1>
    <div
      class="card"
    >
      <div
        class="card-body fill-plain"
      >
        <div
          class="form"
        >
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Course ID:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                CS101
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Course name:
            </label>
            <div
              class="col-sm-6"
              id="coursename"
            >
              <p
                class="form-control-static"
              >
                Introduction to CS
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Sections:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                1
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Teams:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                1
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Total students:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                1
              </p>
            </div>
          </div>
        </div>
        <div
          class="form-group row"
        >
          <label
            class="col-sm-3 control-label text-right"
          >
            Instructors:
          </label>
          <div
            class="col-sm-6"
          >
            
            <div>
               Co-owner: Bran (bran@gmail.com) 
            </div>
          </div>
        </div>
        
        <div>
          <div
            class="form-group"
          >
            <div
              class="text-center"
            >
              <button
                class="btn btn-primary"
                type="button"
                value="Remind Students to Join"
              >
                <i
                  class="fa fa-envelope"
                />
                 Remind Students to Join 
              </button>
              <button
                class="btn btn-primary"
                type="button"
              >
                <i
                  class="fa fa-download"
                />
                 Download Student List 
              </button>
              <button
                class="btn btn-danger"
                type="button"
                value="Delete All Students"
              >
                <i
                  class="fa fa-trash"
                />
                 Delete All Students 
              </button>
              <div>
                <span>
                  Non-English characters not displayed properly in the downloaded file? 
                  <span
                    class="btn-link"
                  >
                     Click here. 
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    
    <div
      class="card-body p-0"
    >
      <div
        class="card-header text-dark-blue bg-light-blue font-weight-bold"
      >
        [CS101]
      </div>
      <div
        class="card-body p-0"
      >
        <tm-student-list
          ng-reflect-course-id="CS101"
          ng-reflect-enable-remind-button="true"
          ng-reflect-sections="[object Object]"
          ng-reflect-use-gray-heading="true"
        />
      </div>
    </div>
    
    
    
    
  </div>
</tm-instructor-course-details-page>
`;

exports[`InstructorCourseDetailsPageComponent should snap with default fields 1`] = `
<tm-instructor-course-details-page
  clipboardService={[Function ClipboardService]}
  courseDetails={[Function Object]}
  courseService={[Function CourseService]}
  courseStudentListAsCsv=""
  httpRequestService={[Function HttpRequestService]}
  instructors={[Function Array]}
  isAjaxSuccess={[Function Boolean]}
  loading="false"
  navigationService={[Function NavigationService]}
  ngbModal={[Function NgbModal]}
  route={[Function ActivatedRoute]}
  router={[Function Router]}
  sections={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
>
  <div>
    <h1>
      Course Details
    </h1>
    <div
      class="card"
    >
      <div
        class="card-body fill-plain"
      >
        <div
          class="form"
        >
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Course ID:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Course name:
            </label>
            <div
              class="col-sm-6"
              id="coursename"
            >
              <p
                class="form-control-static"
              >
                
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Sections:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                0
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Teams:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                0
              </p>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="col-sm-3 text-right"
            >
              Total students:
            </label>
            <div
              class="col-sm-6"
            >
              <p
                class="form-control-static"
              >
                0
              </p>
            </div>
          </div>
        </div>
        <div
          class="form-group row"
        >
          <label
            class="col-sm-3 control-label text-right"
          >
            Instructors:
          </label>
          <div
            class="col-sm-6"
          >
            
          </div>
        </div>
        
      </div>
    </div>
    
    <div
      class="alert alert-warning"
    >
      <i
        class="fa fa-exclamation-circle text-muted"
      />
       There are no students in this course. Click 
      <a
        href="/web/instructor/courses/enroll?courseid="
        ng-reflect-query-params="[object Object]"
        ng-reflect-router-link="/web/instructor/courses/enroll"
        routerlink="/web/instructor/courses/enroll"
      >
        here
      </a>
       to enroll students. 
    </div>
    
    
    
    
  </div>
</tm-instructor-course-details-page>
`;
